Streamlined "My Sites" nav drawer

lynnbgriffith opened the pull request:
#### Changes proposed in this Pull Request

* New hierarchical "My Sites" nav drawer/sidebar structure.
* Move expandable sidebar menu component from reader sidebar to layout components, so it can be reused on other sidebars.
* Implement feature flag to allow us to turn this off per-environment while we test it and slowly roll it out to more users.

_Still in progress:_ Based upon recent design updates, we'll be adding Material icons to the top-level entries on this menu before launching.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I've attempted to keep all the various conditionals and environment-specific tweaks present in this new version of the nav.  For example, "Store" should be hidden when appropriate and custom post types should be included in the "Site" menu.

<img width="1278" alt="Screen Shot 2019-03-25 at 4 15 25 PM" src="https://user-images.githubusercontent.com/449804/54954727-50ce8800-4f19-11e9-8dfd-dfdeed46e79c.png">

lynnbgriffith committed:
Add feature flag for streamlined nav drawer.

lynnbgriffith committed:
Add feature flag for streamlined nav drawer.

lynnbgriffith committed:
Move expandable sidebar menu from Reader so it can be easily reused in other nav drawers.

lynnbgriffith committed:
Move expandable sidebar menu from Reader so it can be easily reused in other nav drawers.

lynnbgriffith committed:
Add a/b test for new nav drawer redesign.

lynnbgriffith committed:
Add a/b test for new nav drawer redesign.

lynnbgriffith committed:
Styles for the expandable sidebar menu component.'

lynnbgriffith committed:
Styles for the expandable sidebar menu component.'

lynnbgriffith committed:
Design menu.

lynnbgriffith committed:
Design menu.

lynnbgriffith committed:
Redux actions, selectors and reducers for My Sites sidebar.

lynnbgriffith committed:
Redux actions, selectors and reducers for My Sites sidebar.

lynnbgriffith committed:
Connect state to expandable sidebar menu components.

lynnbgriffith committed:
Connect state to expandable sidebar menu components.

lynnbgriffith committed:
Tools menu.

lynnbgriffith committed:
Tools menu.

davemart-in commented:
A couple of minor things:

- [ ] @griffbrad let me know if you need any help adding in the material icons for the primary links.

- [ ] @griffbrad when stats are flat, it looks like there is a horizontal line here by mistake: https://cl.ly/987374c4dac8 Note: if this takes longer than a couple of minutes to remedy, we can totally punt it to https://trello.com/b/IWxH1T7j/fixing-the-flows-dotcom-design-march-2019

- [ ] @alexislloyd We appear to be using a blue/grey color for the active state (Stats) and a flat grey color for on hover: https://cl.ly/95d11125b51a These two colors seem too similar. Should that active state color be changed?

- [ ] @griffbrad for the feedbacks link (and any other link that opens in a new tab) we should probably maintain the old external link icon: https://cl.ly/7fef3cece2a4 Otherwise it's too disorienting when you are taken to a new tab. @alexislloyd do you agree/disagree?

- [ ] @allancole thoughts on adding a little extra margin or padding to the bottom of the nav so that as you scroll down, the browser address info toaster message doesn't sit half way over the "Manage" link: https://cl.ly/beedbe848ae2?

- [ ] @allancole notice how when you click to expand a section that the height/padding around the section header increases by a pixel or two: https://cl.ly/ca382e1ddd16 Can we fix this so that the height stays the same whether it's open or closed?

- [ ] @allancole Once Brad adds the Material icons in, we'll probably want to match the font-size and font-weight between the Stats and Plan link text and that of the section headers (which currently are more bold: https://cl.ly/3d4c32c2883b

- [ ] @alexislloyd I could go either way, but thoughts on only allowing one section to be open at a time? Pros: the menu options would never go below the fold. Cons: You may want to reference links in two sections at once (and can't really think of a real-life scenario where this would be the case, but making this change would make this difficult).

- [ ] @griffbrad click on "Plan" then tab and hit enter. The "pages" screen is shown and you lose focus. We'll want to add tabindex on the section title H2's in order to grab focus. We'll also want to make sure that you can toggle the submenus open and closed with both the enter key and the space bar.

jsnajdr commented:
Testing the streamlined sidebar on calypso.live, I can't navigate it with keyboard. On the old version, the sidebar items are focusable, show a focus ring, and can be navigated with Tab and Shift-Tab.

> thoughts on only allowing one section to be open at a time?

How would that affect the accessibility and keyboard-navigability of the sidebar? I have a hunch that the effect might be negative.

drw158 commented:
@davemart-in 

> We appear to be using a blue/grey color for the active state (Stats) and a flat grey color for on hover

We can't really change the saturation or hue of the blue color unless we change the Primary color (which is [Blue 500](https://automattic.github.io/color-studio/)).

> I could go either way, but thoughts on only allowing one section to be open at a time? 

Letting the user control which ones are open would make it more usable. From the gov.au site:

>Accordion elements on some websites collapse any open panels when another one is opened. This behaviour can seem unpredictable for some users, which is why our accordion component does not behave in that way.

>The intent is to make sure that users have full control over what content remains visible to minimise possible confusion when content suddenly ‘disappears’ from view.

Letting multiple sections open at once would allow the customer to "customize" their workspace. It would persist across pageloads, reducing the time and clicks required to get to the link they want.

griffbrad commented:
Just noting here that I don't have @allancole's style work fully integrated quite yet, so will pass it over to him for review once that is in place alongside the Material icons later today.

alexislloyd commented:
>  @alexislloyd I could go either way, but thoughts on only allowing one section to be open at a time? Pros: the menu options would never go below the fold. Cons: You may want to reference links in two sections at once (and can't really think of a real-life scenario where this would be the case, but making this change would make this difficult).

I like the idea of only one section being open at a time. It keeps the view simple. If we run into unexpected issues with it, it's always easy to change, but let's start with that constraint.

> Still in progress: Based upon recent design updates, we'll be adding Material icons to the top-level entries on this menu before launching.

Right now, it's a bit hard to distinguish top-level vs. secondary menu items when a menu is expanded, but I think the addition of icons at the top level will fix this issue.

jsnajdr commented:
> when stats are flat, it looks like there is a horizontal line here by mistake:

This little icon-sized stats chart is called `StatsSparkline` and we discussed it some time ago with @timmyc who did the original implementation. Back then, I was investigating Calypso initial load performance and wondering if we could avoid the HTTP requests this component is issuing on each load.

@timmyc said:
> part of me would say just remove the sparkline altogether, but the history on that feature being added in to calypso was due to a pretty big out cry of users in the forums when old stats was finally retired

and:
> i’d honestly be curious to remove it and see if any users even notice

but:
> my only word of warning with stats is users are crazy about them - someone even made a tamper monkey script to add the sparkline graphic back to the masterbar area

That makes me think: maybe now is finally the right time to remove the `StatsSparkline` from the sidebar?

blowery commented on client/state/action-types.js:
```diff
@@ -543,7 +543,10 @@ export const MAILCHIMP_SETTINGS_LIST = 'MAILCHIMP_SETTINGS_LIST';
 export const MAILCHIMP_SETTINGS_UPDATE_SUCCESS = 'MAILCHIMP_SETTINGS_UPDATE_SUCCESS';
 export const MAILCHIMP_SETTINGS_UPDATE_FAILURE = 'MAILCHIMP_SETTINGS_UPDATE_FAILURE';
-
+export const MY_SITES_SIDEBAR_SITE_TOGGLE = 'MY_SITES_SIDEBAR_SITE_TOGGLE';
```
do we really need one action per menu item?

blowery replied:
though i guess the alternative would be MY_SITES_SIDEBAR_TOGGLE and pass along a const or symbol for each thing in the sidebar... so maybe it's not really much of a win to consolidate

blowery commented on client/state/ui/my-sites/sidebar/reducer.js:
```diff
@@ -0,0 +1,57 @@
+import { combineReducers } from 'state/utils';
+
+export function isSiteOpen( state = false, action ) {
+	switch ( action.type ) {
```
these could all use a common toggleReducer that takes an action to respond to and toggles the state

davemart-in commented:
> We can't really change the saturation or hue of the blue color unless we change the Primary color (which is Blue 500).

@drw158 The current pairing doesn't really go together. Do you disagree? It seems like either:

A) The active style color should be switched from the blue scale to the grey scale, or
B) The hover state style should be moved from the grey scale to the blue scale

alexislloyd commented:
> Letting the user control which ones are open would make it more usable. From the gov.au site:
> 
> Accordion elements on some websites collapse any open panels when another one is opened. This behaviour can seem unpredictable for some users, which is why our accordion component does not behave in that way.
> 
> The intent is to make sure that users have full control over what content remains visible to minimise possible confusion when content suddenly ‘disappears’ from view.
> 
> Letting multiple sections open at once would allow the customer to "customize" their workspace. It would persist across pageloads, reducing the time and clicks required to get to the link they want.

Reading @drw158's comment, I can see the argument for greater usability / a11y here and am on board with that approach.

drw158 commented:
@davemart-in Even though the colors are close, I think it's still helpful to have different colors for the hover and active states. Also, the other color schemes that haven't been launched have more contrasting colors. I'd like to keep the semantic relationship there regardless of the hues we use — `primary` for active and `neutral` for hover. 

The 2 solutions I can think of are either change `primary` to use hot blue (which would be a big change) or add some other visual indicator, like adding a left border to the active state. Another option is to not change anything :)

timmyc commented:
@jsnajdr I still totally agree with my statements from our last discussion on this topic

> That makes me think: maybe now is finally the right time to remove the StatsSparkline from the sidebar?

If this is going to happen, we just need to make sure that HE's are in the loop, specifically `+teamchiron` as much of the user feedback came in the forums. I believe there is even a tampermonkey script that was shared in the forums that added the sparkline back to the masterbar.

Curious - are the changes proposed in this PR also going to be updated so the masterbars/nav shipped elsewhere ( i.e. Claypsoify ) are updated accordingly?

davemart-in commented:
> are the changes proposed in this PR also going to be updated so the masterbars/nav shipped elsewhere ( i.e. Claypsoify ) are updated accordingly?

@timmyc separate PR's, but yes. Here's the one for WP.com. I'll be digging into the code for JP tomorrow.

supernovia commented:
>> my only word of warning with stats is users are crazy about them - someone even made a tamper monkey script to add the sparkline graphic back to the masterbar area

> That makes me think: maybe now is finally the right time to remove the `StatsSparkline` from the sidebar?

Would we consider disabling by default but allowing users to re-enable sparklines for their site?

I know that's more complex, but I was primarily the one answering all the threads about it missing and I can promise, people would notice and be quite unhappy about it. We have some diehard fans, and if we don't *need* to take this away from them, I hope we won't.

lynnbgriffith committed:
Continue integrating new CSS.

lynnbgriffith committed:
Continue integrating new CSS.

lynnbgriffith committed:
Continue integrating new CSS.

lynnbgriffith committed:
Continue integrating new CSS.

lynnbgriffith committed:
Hide sidebar add buttons.

lynnbgriffith committed:
Hide sidebar add buttons.

lynnbgriffith committed:
WP admin link and styles.

lynnbgriffith committed:
WP admin link and styles.

lynnbgriffith committed:
Style tidy.

lynnbgriffith committed:
Style tidy.

lynnbgriffith committed:
Temporarily add some gridicons. Will switch to Material once styles are final.

lynnbgriffith committed:
Temporarily add some gridicons. Will switch to Material once styles are final.

griffbrad commented:
Made progress on the styles.  Temporarily dropped some gridicons into place to better illustrate the state of things as I work out the approach for integrating the Material icons.  Also get the WP Admin added back into the mix with the nice flexbox approach from @allancole that handles window resizing well.

<img width="1198" alt="Screen Shot 2019-03-26 at 4 30 30 PM" src="https://user-images.githubusercontent.com/449804/55034858-7d9ba180-4fe4-11e9-95a0-46d3c11b06ac.png">

Remaining tasks on my list next:

* Redux tidying @blowery suggested.
* Material icons.
* Ensure menu toggle state is handled correctly after a full page reload.
* tabindex fixes.

lynnbgriffith committed:
Correct vertical positioning of external link icon.

lynnbgriffith committed:
Correct vertical positioning of external link icon.

Commit:
Merge branch 'master' into try/streamlined-nav-drawer

Commit:
Fix hover->active 1px bug

See: https://github.com/Automattic/wp-calypso/issues/31774

Before: https://cl.ly/17bed1707473
After: https://cl.ly/ac3a117a57f2

Commit:
Fix hover->active 1px bug

See: https://github.com/Automattic/wp-calypso/issues/31774

Before: https://cl.ly/17bed1707473
After: https://cl.ly/ac3a117a57f2

sixhours commented:
👋How is the "Add New Site" button handled in this reorg? Are we considering both scenarios -- accounts with just one site and accounts with multiple sites? There's a related ticket in #31636 that I'm hoping this nav drawer reorg will be able to answer/fix. :)

Commit:
Add margin to the bottom of nav drawer

See: https://github.com/Automattic/wp-calypso/issues/31775

Before: https://cl.ly/5b1b4b5f6eab
After: https://cl.ly/2f3ccf3e7c89

Commit:
Add margin to the bottom of nav drawer

See: https://github.com/Automattic/wp-calypso/issues/31775

Before: https://cl.ly/5b1b4b5f6eab
After: https://cl.ly/2f3ccf3e7c89

davemart-in commented:
@sixhours I believe we've opted to leave that one as is for this project. We'll have to address it separately in another PR.

lynnbgriffith committed:
Add material-design-icons package for SVGs we need on the nav drawer.

lynnbgriffith committed:
Add material-design-icons package for SVGs we need on the nav drawer.

lynnbgriffith committed:
Add earn link to tools menu.

lynnbgriffith committed:
Add earn link to tools menu.

lynnbgriffith committed:
Integrate svgr and use it to load SVGs with webpack.

lynnbgriffith committed:
Integrate svgr and use it to load SVGs with webpack.

lynnbgriffith committed:
Material icons in my-sites sidebar and related components.

lynnbgriffith committed:
Material icons in my-sites sidebar and related components.

Commit:
Remove site avatar from .current-section component

Also swapped order of site title and section title and made a few minor
CSS changes. See: https://github.com/Automattic/wp-calypso/issues/31772

Before: https://cl.ly/366181e5e690
After: https://cl.ly/2d8800f199d5

Note: I also addressed some linting issues. See:
https://cl.ly/07dacf96db6c

Commit:
Remove site avatar from .current-section component

Also swapped order of site title and section title and made a few minor
CSS changes. See: https://github.com/Automattic/wp-calypso/issues/31772

Before: https://cl.ly/366181e5e690
After: https://cl.ly/2d8800f199d5

Note: I also addressed some linting issues. See:
https://cl.ly/07dacf96db6c

lynnbgriffith committed:
Update shrinkwrap.

lynnbgriffith committed:
Merge branch 'try/streamlined-nav-drawer' of github.com:Automattic/wp-calypso into try/streamlined-nav-drawer

lynnbgriffith committed:
Update package.json fields that still had placeholders from the docs.

lynnbgriffith committed:
Update package.json fields that still had placeholders from the docs.

lynnbgriffith committed:
Update shrinkwrap with corrected material-design-icons package.

lynnbgriffith committed:
Only allow new users on our a/b test.

lynnbgriffith committed:
Only allow new users on our a/b test.

lynnbgriffith committed:
Add mySites reducer to ui state test.

lynnbgriffith committed:
Add mySites reducer to ui state test.

lynnbgriffith committed:
Reuse reducer logic for menu toggles.

lynnbgriffith committed:
Reuse reducer logic for menu toggles.

lynnbgriffith committed:
Add some info about adding icons to the package.

lynnbgriffith committed:
Add some info about adding icons to the package.

lynnbgriffith committed:
Incorporate changes from PR #31616.

Incorporate @apeatling's changes to the view site links.

This isn't a direct copy of his diff because I didn't want to remove the
"View Site"/preview link in full while we're still in the a/b testing
stage.

lynnbgriffith committed:
Incorporate changes from PR #31616.

Incorporate @apeatling's changes to the view site links.

This isn't a direct copy of his diff because I didn't want to remove the
"View Site"/preview link in full while we're still in the a/b testing
stage.

blowery commented on client/state/ui/reducer.js:
```diff
@@ -23,6 +23,7 @@ import gutenbergOptInDialog from './gutenberg-opt-in-dialog/reducer';
 import language from './language/reducer';
 import layoutFocus from './layout-focus/reducer';
 import mediaModal from './media-modal/reducer';
+import mySites from './my-sites/reducer';
```
we should really pull this in only for routes that load my-sites. @jsnajdr can you help with that?

blowery commented on packages/material-design-icons/src/index.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+	STATIC_PATH: __dirname,
```
is this used by anything?

griffbrad replied:
I pulled that from the official material-design-icons package.  My thought was that it might make it easier for folks to move from that package to this alternative (either in full or just to get SVGs missing from the official package), if they were relying on that for path resolution.

blowery commented:
@griffbrad to fix up the shrinkwrap conflict, you can use the [npm merge driver](https://www.npmjs.com/package/npm-merge-driver)

lynnbgriffith committed:
Merge branch 'try/streamlined-nav-drawer' of github.com:Automattic/wp-calypso into try/streamlined-nav-drawer

lynnbgriffith committed:
Update jest config to support svgr imports in client tests.

lynnbgriffith committed:
Adjust approach for integrating svgr with Jest.

In order to support the flagUrl() library function, we now mock svgr's
return value, including the "default" export with the URL to the SVG
file.  This allows us to continue importing the ReactComponent for an
SVG while also supporting use cases like flagUrl().

lynnbgriffith committed:
Implement keyboard accessibility for expandable sidebar menu components.

lynnbgriffith committed:
Disable eslint check for events on non-interactive elements.

Currently, the toggles for expandable sidebar menus use <h2> elements,
which are not interactive elements.  Ultimately, we need to refactor
these components to use a <button> element.  However, doing so at the
moment would break the current sidebar design (still in the code while
we a/b test the new version) and drastically complicate the styles.

For now, I've kept the <h2> and its mouse and keyboard events.

lynnbgriffith committed:
Implement a/b test check when in prod.

lynnbgriffith committed:
Check a/b test variation before adding outer CSS class.

lynnbgriffith committed:
Check env_id from config prior to evaluating a/b test variation.

lynnbgriffith committed:
Re-add CSS class name Woo e2e tests rely upon.

lynnbgriffith committed:
Add a Material icon for store menu item.

lynnbgriffith committed:
Incorporate package.json changes from master.

lynnbgriffith committed:
Add "Activity" to Tools menu.

lynnbgriffith committed:
Switch back to actual Jetpack logo and drop PlanSvg.

lynnbgriffith committed:
Re-order Design menu items to reflect expected frequency of use.

lynnbgriffith committed:
Minor style tweaks based upon recent feedback.

lynnbgriffith committed:
Use color variable to top border of wp-admin footer.

Commit:
Remove excess margin in responsive web view

Before: https://cl.ly/f2c590a73d7f
After: https://cl.ly/f2898f4257e5

Commit:
Fixing horizontal borders in responsive web view

Before: https://cl.ly/3d493f3f0fc2
After: https://cl.ly/1db2a8515b7b

griffbrad commented:
This PR has been superseded by PR #32370.
